### DISCUSSION TITLE
Schema for representing instantaneous states (of satellites and aerial vehicles)

### DISCUSSION THREAD

**Discussion by vinay0000** [eose-tools-org/eose-api#5] [General]
The results of orbit propagation over a mission duration would be a time series of state of the satellite/ aerial-vehicles

_State representations in relevant software:_

[STK Coordinate type](https://help.agi.com/stk/#stk/vehSat_coordType.htm): STK calls it as 'coordinate type' for 'vehicles', where each 'coordinate type is a unique representation of the position and velocity of a vehicle.'

_Two potential representations could be using Keplerian elements or Cartesian states.._

**Instantaneous Keplerian elements**  (for satellites only)
Related STK implementation is [here](https://help.agi.com/stk/#stk/vehSat_coordType_classical.htm). Our implementation could be restricted to one particular set of Keplerian elements for simplicity.

Example implementation:
```
 {  "type": "osculating_keplerian_elements",
       "date": { ... },
	"state": {
    "semimajor_axis": 10000,
    "eccentricity": 0.1,
    "inclination": 30,
    "argument_of_perigee": 60,
    "right_ascension_of_ascending_node": 45,
    "true_anomaly": 20
  },
  "reference_frame": "J2000"
 }
```

**Instantaneous Cartesian State Vector**: (for both satellites and aerial vehicles) 

Example implementation:
```
{
  "type": "cartesian_state_vector",
  "date": {...},
  "state": {
    "x": 8000,
    "y": 2000,
    "z": 3000,
    "vx": 7,
    "vy": 0.2,
    "vz": 0.4
  },
   "reference_frame": "J2000"
}

```
The reference frame could be limited to J2000 for the Keplerian representation, and J2000 and an appropriate ECEF reference frame (ITRF?) for the Cartesian representation. ECEF frame would be more appropriate in case of aerial vehicle applications.

**Representing time-series of states**

- One potential representation is of the states as a list, where each element of the list is the state vector with corresponding time-stamp.
- Commonly the states would be available at a fixed cadence (time-period). In such as case an 'epoch' ``date `` object with a``time_step`` key can be specified. A ``time_index`` key is specified with each state, and the absolute time corresponding to the state can be calculated as: ``epoch + time_step * time_index``

Example implementation:
```
{
  "type": "cartesian_state_vector",
  "epoch": {...},
  "time_step": 60,
[ 
  "state": {
    "x": 8000,
    "y": 2000,
    "z": 3000,
    "vx": 7,
    "vy": 0.2,
    "vz": 0.4,
    "time_index": 0
  },
 "state": {
    "x": 8000,
    "y": 2500,
    "z": 3000,
    "vx": 7,
    "vy": 0.2,
    "vz": 0.4,
    "time_index": 1
  },
...
],
   "reference_frame": "J2000"
}
```

**_Other Notes_**
The discussion on ``date`` object schema is [here](https://github.com/eose-tools-org/eose-api/discussions/3)

Some relevant copied bullets about the J2000 reference frame and its relation to the ICRF frame, from the [presentation](https://naif.jpl.nasa.gov/pub/naif/toolkit_docs/Tutorials/pdf/individual_docs/17_frames_and_coordinate_systems.pdf) in the context of [JPL SPICE Toolkit](https://naif.jpl.nasa.gov/naif/toolkit.html):
> - The realization of ICRF was made to coincide almost exactly with the J2000 frame.
> - The difference is very small–a rotation of less than 0.1 arc second.
> - The reference frame name "J2000" is generally used in SPICE as a label for the ICRF frame.

**Comment by ryanketzner** [comment]
Something that could help guide discussion: what are everybody's needs for state representation for their current tools? For now, Vinay and I use the "time series of states" representation as input to our coverage calculator. That is, the coverage calculator requires only discrete points along the trajectory as input and does not need to know the entire trajectory or the model used to generate it. 

I think some people's software will require a representation that gives the state as an arbitrary function of time. Some examples of this type of representation are:

1. SGP4. Given the initial condition and initial time, the SGP4 model tells the downstream software how to compute the state at any other time.
2. J2-perturbed two body problem.
3. [JPL ephemerides](https://en.wikipedia.org/wiki/Jet_Propulsion_Laboratory_Development_Ephemeris), which store the pre-solved trajectory as a function of time using Chebyshev polynomials. 
4. An unperturbed circular orbit, whose trajectory can be given analytically as a function of time by a simple parametric equation.

It should generally be possible to convert these types of representations to the "time series of states" representation, but not the other way around. There is also some distinction to be made between examples (1) and (2), which give a model which can be used by the downstream software to generate the states, and examples (3) and (4), which give a representation of the trajectory as an analytical function of time. 

@jsipps26 you might have some input on this since if I recall correctly you use a continuous coverage simulation.

**Reply by jsipps26** [reply]
I do use a time-series of states for the spacecraft itself, or the states are recovered from the orbit model within coverage calculation. The discrete series acts as an interpolant for the assignment of continuous access time to each grid point. Internally I've made an Orbit class that either uses a prescribed propagator like your 1st or 2nd options, but it can also read in an ephemeris file (e.g., from GMAT), then either interpolate within the file bounds or extrapolate (if necessary) under a default propagator.
